File: c_files/dsa/singly_linked_list.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node
{
    char data;
    struct Node* next;

} Node;

Node*
create_node(char data)
{
    Node* new_node = malloc(sizeof(Node));
    if (new_node == NULL)
    {
        fprintf(stderr, "Memory allocation failed\n");
        exit(EXIT_FAILURE);
    }

    new_node->data = data;
    new_node->next = NULL;
}

typedef struct LinkedList
{
    Node* head;

} LinkedList;

void
ll_init(LinkedList* ll)
{
    ll->head = NULL;
}

int
ll_length(LinkedList* ll)
{
    if (!ll)
    {
        fprintf(stderr, "List not initialized\n");
        exit(EXIT_FAILURE);
    }

    Node* current = ll->head;
    int count = 0;
    while (current)
    {
        count++;
        current = current->next;
    }

    return count;
}

void
ll_prepend(LinkedList* ll, char data)
{
    if (!ll)
    {
        fprintf(stderr, "List not initialized\n");
        exit(EXIT_FAILURE);
    }

    Node* new_node = create_node(data);
    new_node->next = ll->head;
    ll->head = new_node;
    return;
}

void
ll_append(LinkedList* ll, char data)
{
    if (!ll)
    {
        fprintf(stderr, "List not initialized\n");
        exit(EXIT_FAILURE);
    }

    Node* new_node = create_node(data);

    if (!ll->head)
    {
        new_node->next = ll->head;
        ll->head = new_node;
        return;
    }

    Node* current = ll->head;
    while (current->next)
    {
        current = current->next;
    }

    current->next = new_node;
}

void
ll_insert_at_index(LinkedList* ll, int index, char character)
{
    if (!ll)
    {
        fprintf(stderr, "List not initialized\n");
        exit(EXIT_FAILURE);
    }

    if (index < 0)
    {
        fprintf(stderr, "Index %d invalid\n", index);
        return;
    }

    Node* new_node = create_node(character);

    if (index == 0)
    {
        new_node->next = ll->head;
        ll->head = new_node;
        return;
    }

    Node* prev = ll->head;
    int count = 0;
    while (prev && count < index - 1)
    {
        count++;
        prev = prev->next;
    }

    if (!prev)
    {
        fprintf(stderr, "Index %d out of bounds\n", index);
        free(new_node);
        return;
    }

    new_node->next = prev->next;
    prev->next = new_node;
}

void
ll_print_list(LinkedList* ll)
{
    if (!ll)
    {
        fprintf(stderr, "List not initialized\n");
        exit(EXIT_FAILURE);
    }

    Node* current = ll->head;
    while (current)
    {
        printf("%c -> ", current->data);
        current = current->next;
    }
    printf("None\n");
}

void
ll_free(LinkedList* ll)
{
    if (!ll)
    {
        fprintf(stderr, "List not initialized\n");
        exit(EXIT_FAILURE);
    }

    Node* current = ll->head;
    while (current)
    {
        Node* tmp = current->next;
        free(current);
        current = tmp;
    }
}

int
main(void)
{
    LinkedList ll;
    ll_init(&ll);
    ll_append(&ll, 'b');
    ll_append(&ll, 'c');
    ll_append(&ll, 'd');
    ll_prepend(&ll, 'a');
    ll_print_list(&ll);
    printf("Length: %d\n", ll_length(&ll));
    ll_insert_at_index(&ll, 2, 'f');
    ll_print_list(&ll);
    printf("Length: %d\n", ll_length(&ll));

    ll_free(&ll);

    return 0;
}

File: python_files/dsa/singly_linked_list.py
class Node:
    def __init__(self, data: int) -> None:
        self.data: int = data
        self.next: Node | None = None


class LinkedList:
    def __init__(self) -> None:
        self.head: Node | None = None

    def insert_at_beginning(self, data: int) -> None:
        new_node: Node = Node(data)
        new_node.next = self.head
        self.head = new_node

    def insert_at_end(self, data: int) -> None:
        new_node: Node = Node(data)

        if self.head is None:
            self.head = new_node
            return

        current: Node = self.head
        while current.next:
            current = current.next
        current.next = new_node

    def insert_at_index(self, index: int, data: int) -> None:
        new_node: Node = Node(data)

        if index == 0:
            new_node.next = self.head
            self.head = new_node
            return

        current: Node | None = self.head
        current_index: int = 0
        while current:
            current_index += 1
            if current_index == index:
                if current.next is None:
                    current.next = new_node
                    return
                else:
                    prev: Node | None = current
                    new_node.next = current.next
                    prev.next = new_node

            current = current.next

        if index > current_index:
            raise ValueError("Index out of bounds")

    def delete_by_value(self, data: int) -> None:
        if self.head is None:
            print("List is empty")
            return
        elif self.head.data == data:
            self.head = self.head.next
            return

        prev: Node | None = self.head
        while prev:
            current: Node | None = prev.next
            if current is not None and current.data == data:
                prev.next = current.next
                return

            prev = prev.next

    def delete_by_index(self, index: int) -> None:
        if self.head is None:
            print("List is empty")
            return
        elif index == 0:
            self.head = self.head.next
            return

        prev: Node | None = self.head
        current_index: int = 0
        while prev:
            current_index += 1
            current: Node | None = prev.next
            if current is not None and index == current_index:
                prev.next = current.next
                return

            prev = prev.next

    def print(self) -> None:
        if self.head is None:
            print("List is empty")
            return

        current: Node | None = self.head
        while current:
            print(current.data, end=" -> ")
            current = current.next
        print("None")

    def length(self) -> int:
        count: int = 0
        current: Node | None = self.head
        while current:
            count += 1
            current = current.next

        return count


ll: LinkedList = LinkedList()
ll.insert_at_end(1)
ll.insert_at_end(2)
ll.insert_at_end(3)
ll.insert_at_beginning(0)
ll.print()
ll.insert_at_index(0, 2)
ll.print()
